1. Import necessary libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np

2. Import dataset (employee_dataset.csv)

In [2]:
# Import dataset

data = pd.read_csv('employee_dataset.csv')
data.head()

Age     BusinessTravel  MonthlyIncome  JobSatisfaction  Bonus  \
0   41      Travel_Rarely           5993                4  17979   
1   49  Travel_Frequently           5130                2  20520   
2   37      Travel_Rarely           2090                3   6270   
3   33  Travel_Frequently           2909                3   8727   
4   27      Travel_Rarely           3468                2  10404   

               Department  DistanceFromHome  Education EducationField  \
0                   Sales                 1          2  Life Sciences   
1  Research & Development                 8          1  Life Sciences   
2  Research & Development                 2          2          Other   
3  Research & Development                 3          4  Life Sciences   
4  Research & Development                 2          1        Medical   

   EmployeeCount  ...                JobRole  MaritalStatus PerformanceRating  \
0              1  ...        Sales Executive         Single                 3   
1              1  ...     Research Scientist        Married                 4   
2              1  ...  Laboratory Technician         Single                 3   
3              1  ...     Research Scientist        Married                 3   
4              1  ...  Laboratory Technician        Married                 3   

   StockOptionLevel TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                 0                     0               1               6   
1                 1                     3               3              10   
2                 0                     3               3               0   
3                 0                     3               3               8   
4                 1                     3               3               2   

   YearsSinceLastPromotion  OverTime  Attrition  
0                        0       Yes        Yes  
1                        1        No         No  
2                        0       Yes        Yes  
3                        3       Yes         No  
4                        2        No         No  

[5 rows x 24 columns]

In [3]:
# Show data info

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Age                      1470 non-null   int64 
 1   BusinessTravel           1470 non-null   object
 2   MonthlyIncome            1470 non-null   int64 
 3   JobSatisfaction          1470 non-null   int64 
 4   Bonus                    1470 non-null   int64 
 5   Department               1470 non-null   object
 6   DistanceFromHome         1470 non-null   int64 
 7   Education                1470 non-null   int64 
 8   EducationField           1470 non-null   object
 9   EmployeeCount            1470 non-null   int64 
 10  EmployeeNumber           1470 non-null   int64 
 11  EnvironmentSatisfaction  1470 non-null   int64 
 12  Gender                   1470 non-null   object
 13  JobLevel                 1470 non-null   int64 
 14  JobRole                  1470 non-null  

In [4]:
# Drop other column and keep 'Age', 'BusinessTravel', 'Monthly Income', 'JobSatisfaction' and 'Attrition'

data = data.iloc[:, [0, 1, 2, 3, -1]]
data.head()

Age     BusinessTravel  MonthlyIncome  JobSatisfaction Attrition
0   41      Travel_Rarely           5993                4       Yes
1   49  Travel_Frequently           5130                2        No
2   37      Travel_Rarely           2090                3       Yes
3   33  Travel_Frequently           2909                3        No
4   27      Travel_Rarely           3468                2        No

In [5]:
# Show data shape

data.shape

(1470, 5)

In [6]:
# Check missing data

data.isnull().values.any()

False

In [7]:
# Describe data

data.describe()

Age  MonthlyIncome  JobSatisfaction
count  1470.000000    1470.000000      1470.000000
mean     36.923810    6502.931293         2.728571
std       9.135373    4707.956783         1.102846
min      18.000000    1009.000000         1.000000
25%      30.000000    2911.000000         2.000000
50%      36.000000    4919.000000         3.000000
75%      43.000000    8379.000000         4.000000
max      60.000000   19999.000000         4.000000

3. Allocate the relevant attributes as input and output

In [8]:
# Allocate the relevant attributes as input (X) and output (y)

X = data[['Age', 'BusinessTravel', 'MonthlyIncome', 'JobSatisfaction']]
y = data['Attrition']

4. Use LabelEncoder to encode categorical data

In [9]:
# Use LabelEncoder to encode categorical data

from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()
X['BusinessTravel'] = label_encoder.fit_transform(X['BusinessTravel'])
print(X.head())

   Age  BusinessTravel  MonthlyIncome  JobSatisfaction
0   41               2           5993                4
1   49               1           5130                2
2   37               2           2090                3
3   33               1           2909                3
4   27               2           3468                2


C:\Users\mohdz\AppData\Local\Temp\ipykernel_21824\2278383336.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['BusinessTravel'] = label_encoder.fit_transform(X['BusinessTravel'])


5. Split your data into training and test sets with the appropriate proportions

In [10]:
# Split data into training and test sets with the appropriate proportions

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Training set shape:", X_train.shape, y_train.shape)
print("Test set shape:", X_test.shape, y_test.shape)

Training set shape: (1176, 4) (1176,)
Test set shape: (294, 4) (294,)


6. Normalized your data using StandardScaler

In [11]:
# Normalized data using StandardScaler

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Scaled Training Set:")
print(X_train_scaled[:5])

Scaled Training Set:
[[-1.38855944  0.58928051 -0.86827746 -1.58233648]
 [-2.04073779 -2.46355596 -1.13280679  1.15283407]
 [-0.84507748  0.58928051  0.11653658  1.15283407]
 [ 0.24188644  0.58928051 -0.23106984 -0.67061296]
 [-0.62768469  0.58928051 -0.46436559  0.24111056]]


7. Fit the and predict results using the SVM Classifier

In [12]:
# Fit the and predict results using the SVM Classifier

from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

svm_classifier = SVC()
svm_classifier.fit(X_train_scaled, y_train)
y_pred = svm_classifier.predict(X_test_scaled)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

Accuracy: 0.87


8. Fit the and predict results using the Naïve Bayes Classifier

In [13]:
# Fit the and predict results using the Naïve Bayes Classifier

from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

nb_classifier = GaussianNB()
nb_classifier.fit(X_train_scaled, y_train)
y_pred = nb_classifier.predict(X_test_scaled)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

Accuracy: 0.87


9. Fit the and predict results using the ANN Classifier

In [14]:
# Fit the and predict results using the ANN Classifier

from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

ann_classifier = MLPClassifier(hidden_layer_sizes=(8,), max_iter=50, random_state=42)
ann_classifier.fit(X_train, y_train)
y_pred = ann_classifier.predict(X_test_scaled)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

Accuracy: 0.45


C:\Users\mohdz\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (50) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\mohdz\anaconda3\Lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


10. Fit the and predict results using the Decision Tree Classifier

In [15]:
# Fit the and predict results using the Decision Tree Classifier

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

dt_classifier = DecisionTreeClassifier(random_state=42)
dt_classifier.fit(X_train_scaled, y_train)
y_pred = dt_classifier.predict(X_test_scaled)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

Accuracy: 0.76


11. Evaluate your results using confusion matrix and calculate the prediction accuracy for both models

In [16]:
# Evaluate results using confusion matrix

from sklearn.metrics import confusion_matrix, accuracy_score

# Evaluate SVM Classifier

y_pred_svm = svm_classifier.predict(X_test_scaled)
accuracy_svm = accuracy_score(y_test, y_pred_svm)
conf_matrix_svm = confusion_matrix(y_test, y_pred_svm)
print("SVM Classifier:")
print(f"Accuracy: {accuracy_svm:.2f}")
print("Confusion Matrix:")
print(conf_matrix_svm)

# Evaluate Naïve Bayes Classifier

y_pred_nb = nb_classifier.predict(X_test_scaled)
accuracy_nb = accuracy_score(y_test, y_pred_nb)
conf_matrix_nb = confusion_matrix(y_test, y_pred_nb)
print("\nNaïve Bayes Classifier:")
print(f"Accuracy: {accuracy_nb:.2f}")
print("Confusion Matrix:")
print(conf_matrix_nb)

#Evaluate ANN Classifier

y_pred_ann = ann_classifier.predict(X_test_scaled)
accuracy_ann = accuracy_score(y_test, y_pred_ann)
conf_matrix_ann = confusion_matrix(y_test, y_pred_ann)
print("\nANN Classifier:")
print(f"Accuracy: {accuracy_ann:.2f}")
print("Confusion Matrix:")
print(conf_matrix_ann)

# Evaluate Decision Tree Classifier

y_pred_dt = dt_classifier.predict(X_test_scaled)
accuracy_dt = accuracy_score(y_test, y_pred_dt)
conf_matrix_dt = confusion_matrix(y_test, y_pred_dt)
print("\nDecision Tree Classifier:")
print(f"Accuracy: {accuracy_dt:.2f}")
print("Confusion Matrix:")
print(conf_matrix_dt)

SVM Classifier:
Accuracy: 0.87
Confusion Matrix:
[[255   0]
 [ 39   0]]

Naïve Bayes Classifier:
Accuracy: 0.87
Confusion Matrix:
[[255   0]
 [ 38   1]]

ANN Classifier:
Accuracy: 0.45
Confusion Matrix:
[[119 136]
 [ 25  14]]

Decision Tree Classifier:
Accuracy: 0.76
Confusion Matrix:
[[214  41]
 [ 31   8]]


C:\Users\mohdz\anaconda3\Lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


12. Discuss your results and findings.

Both the SVM and Naïve Bayes classifiers achieved an accuracy of 87%. However, their performance differs slightly in terms of the confusion matrix. The SVM classifier correctly classified all instances of the negative class (no attrition) but misclassified all instances of the positive class (attrition), resulting in many false negatives. On the other hand, the Naïve Bayes classifier also correctly classified all instances of the negative class but correctly predicted one instance of the positive class, resulting in a lower number of false negatives.

The Artificial Neural Network (ANN) classifier performed poorly compared to the SVM and Naïve Bayes classifiers, with an accuracy of only 45%. The confusion matrix reveals many false negatives and false positives, indicating a high degree of misclassification.

The Decision Tree classifier achieved an accuracy of 76%. While its accuracy is lower than that of the SVM and Naïve Bayes classifiers, it performs better than that of the ANN classifier. The confusion matrix shows a moderate number of false negatives and false positives, indicating some degree of misclassification but not as severe as that of the ANN classifier.

13. Based on your findings, propose a model to be used at the company to predict the employee attrition.

Naïve Bayes classifier. The Naïve Bayes classifier achieved a high accuracy of 87%, indicating its effectiveness in predicting employee attrition. The confusion matrix shows that the Naïve Bayes classifier correctly predicted a significant portion of the positive and negative cases of attrition, with only a few false predictions. Naïve Bayes classifiers provide transparent and interpretable results, making understanding the factors influencing attrition predictions easy. The Naïve Bayes classifier naturally handles categorical features well, which is often the case in HR datasets where factors like job role, department, and business travel frequency are categorical. This interpretability can be valuable for HR professionals when making informed decisions. The Naïve Bayes classifier appears to be suitable for predicting employee attrition at the company. Its high accuracy, simplicity, ability to handle categorical features, and interpretability make it a reliable model for this task.